src/ai/backend/agent/agent.py

```diff
@@ -1701,6 +1704,80 @@ async def create_kernel(
                             affinity_hint = AffinityHint(
                                 hint_devices, self.affinity_map, affinity_hint.policy
                             )
+                        except InsufficientResource as e:
+                            await self.produce_event(
+                                DoAgentResourceCheckEvent(
+                                    ctx.agent_id,
+                                )
+                            )
+                            plugin_alloc_map: MutableMapping[
+                                DeviceName,
+                                MutableMapping[SlotName, MutableMapping[DeviceId, Decimal]],
+                            ] = {
+                                dev: dict(self.computers[dev].alloc_map.allocations)
+                                for dev in ordered_dev_names
+                            }
+                            original_plugin_alloc_map = copy.deepcopy(plugin_alloc_map)
+                            for device_name, device_dict in plugin_alloc_map.items():
```

Comment:
How about just iterate allocations from the exception? I think that we do not need to iterate all the computer devices and copy all of them.
And if you want to iterate all computer devices here, it would be better to iterate the devices and set those key and value into `plugin_alloc_map` and `original_plugin_alloc_map` one by one rather than using list comprehension and copy them because we need at least two iterations here and `deepcopy()` causes additional recursions that can be slower than normal iterations.

Reply:
This code attempts increment while assigning key and value one by one without using deepcopy.
```python
plugin_alloc_map: dict[
    DeviceName, dict[SlotName, MutableMapping[DeviceId, Decimal]]
] = {}
original_plugin_alloc_map: dict[
    DeviceName, dict[SlotName, MutableMapping[DeviceId, Decimal]]
] = {}
for dev in ordered_dev_names:
    plugin_alloc_map[dev] = dict(
        self.computers[dev].alloc_map.allocations
    )
    original_plugin_alloc_map[dev] = dict(
        self.computers[dev].alloc_map.allocations
    )
for temp_slot_name, temp_slot_dict in e.args[5].items():
    print(temp_slot_name)
    print(temp_slot_dict)
    print("------------")
    for device_id in temp_slot_dict.keys():
        if (
            device_id
            in plugin_alloc_map[dev_name][temp_slot_name].keys()
        ):
            print(
                "id(plugin_alloc_map[dev_name][temp_slot_name]): ",
                id(plugin_alloc_map[dev_name][temp_slot_name]),
            )
            print(
                "id(original_plugin_alloc_map[dev_name][temp_slot_name]): ",
                id(original_plugin_alloc_map[dev_name][temp_slot_name]),
            )
            print(
                "id(plugin_alloc_map[dev_name][temp_slot_name][device_id]): ",
                id(
                    plugin_alloc_map[dev_name][temp_slot_name][
                        device_id
                    ]
                ),
            )
            print(
                "id(original_plugin_alloc_map[dev_name][temp_slot_name][device_id]): ",
                id(
                    original_plugin_alloc_map[dev_name][temp_slot_name][
                        device_id
                    ]
                ),
            )
            print("----increment processing----")
            plugin_alloc_map[dev_name][temp_slot_name][
                device_id
            ] += temp_slot_dict[device_id]
            print(
                "id(plugin_alloc_map[dev_name][temp_slot_name]): ",
                id(plugin_alloc_map[dev_name][temp_slot_name]),
            )
            print(
                "id(original_plugin_alloc_map[dev_name][temp_slot_name]): ",
                id(original_plugin_alloc_map[dev_name][temp_slot_name]),
            )
            print(
                "id(plugin_alloc_map[dev_name][temp_slot_name][device_id]): ",
                id(
                    plugin_alloc_map[dev_name][temp_slot_name][
                        device_id
                    ]
                ),
            )
            print(
                "id(original_plugin_alloc_map[dev_name][temp_slot_name][device_id]): ",
                id(
                    original_plugin_alloc_map[dev_name][temp_slot_name][
                        device_id
                    ]
                ),
            )
            print("----------")
            print("original_plugin_alloc_map: ")
            pprint.pprint(original_plugin_alloc_map)
            print("plugin_alloc_map: ")
            pprint.pprint(plugin_alloc_map)
```
print results:
<img width="696" alt="스크린샷 2023-08-15 01 57 12" src="https://github.com/lablup/backend.ai/assets/59244452/b7ac8619-5aa7-4f31-8ef2-f51b26aecd1a">

It's example that allocate `{'cpu': {'2': Decimal('1')}`.
As you can see, `id(plugin_alloc_map[dev_name][temp_slot_name])` == `id(original_plugin_alloc_map[dev_name][temp_slot_name])`.
The id according to the device name is different because it is assigned according to the device name key, but the slot name points to the same id. (because it's nested dict)
So after increment, we can see that `original_plugin_alloc_map` and `plugin_alloc_map` have the same value.

This is the result when using deep copy.
<img width="696" alt="스크린샷 2023-08-15 02 15 45" src="https://github.com/lablup/backend.ai/assets/59244452/5e509478-e848-4959-abf1-d895969318a2">
As you can see, `id(plugin_alloc_map[dev_name][temp_slot_name])` != `id(original_plugin_alloc_map[dev_name][temp_slot_name])`.
So after increment, we can see that `original_plugin_alloc_map` and `plugin_alloc_map` have different value. (cpu, device_id == 2)

So I decided to use deepcopy.
Please let me know if there is any other option!

Reply:
Thanks!
Modified the code to use the least amount of deepcopy().
And just iterate allocations from the exception to efficiently construct the for loop.
plz check 👍